### PR TITLE
AO3-6629 Switch to signed JSON-formatted session cookies

### PR DESCRIPTION
**Pull request by AlbertSPedersen**
# Pull Request Checklist

* [X] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [X] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [X] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [X] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [X] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6629

## Purpose

This switches from encrypted to signed-only session cookies and changes the format to JSON for easy parsing. This enables session inspection by an intermediary proxy such as NGINX or Cloudflare.

## Credit

Albert Pedersen (he/him)

### CODE REVIEW

File: config/initializers/cookies_serializer.rb [hunk 1]
```diff
@@ -2,4 +2,4 @@
 
 # Specify a serializer for the signed and encrypted cookie jars.
 # Valid options are :json, :marshal, and :hybrid.
-Rails.application.config.action_dispatch.cookies_serializer = :marshal
+Rails.application.config.action_dispatch.cookies_serializer = :hybrid
```
[the remark this answers is not in the document]
Comment (AlbertSPedersen) [reply config/initializers/cookies_serializer.rb:5]: This is a false positive. The cookies are signed and will not be deserialized without a valid signature. The same warning occurs with `:marshal` which was the previous setting: https://github.com/presidentbeef/brakeman/blob/391f530d6d685455fa4a5a8c9f5d29d6bc03af73/lib/brakeman/checks/check_cookie_serialization.rb#L12.

This was [discussed](https://github.com/presidentbeef/brakeman/discussions/1710#discussioncomment-2801601) in the Brakeman project:
> Crafting a malicious session cookie does require the attacker to know the secret key ...

This was set to `:hybrid` rather than `:json` to avoid breaking existing "remember me" cookies.

File: config/initializers/session_store.rb [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
+  # regardless of whether a secret_key_base has been set
+  def cookie_jar(request)
+    request.cookie_jar.signed
+  end
+end
+
 # Be sure to restart your server when you modify this file.
 
-Otwarchive::Application.config.session_store :cookie_store, key: '_otwarchive_session', expire_after: ArchiveConfig.DEFAULT_SESSION_LENGTH_IN_WEEKS.weeks
+Otwarchive::Application.config.session_store :force_signed_cookie_store, key: '_otwarchive_session', expire_after: ArchiveConfig.DEFAULT_SESSION_LENGTH_IN_WEEKS.weeks
```
[the remark this answers is not in the document]
Comment (AlbertSPedersen) [reply config/initializers/session_store.rb:13]: The string in question was not modified.

File: config/initializers/cookies_serializer.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (zz9pzza) [reply config/initializers/cookies_serializer.rb:5]: Will this change in the future, should we have a todo ?

[the remark this answers is not in the document]
Comment (AlbertSPedersen) [reply config/initializers/cookies_serializer.rb:5]: It would be a good idea to change the format to `:json` once all current "remember me" tokens have expired. That will be 3 months after this change is deployed (expiration of `remember_user_token` cookie).

**Comment by TheCarpetMerchant** [comment]
Hello,

If I may bother someone to get confirmation of a testing protocol for this : would creating a session (with remember_me token) without this change, then applying the patch and keep using the session result in equivalent behavior as to what is done for AO3 deployments ?
Just to make sure that what I'm testing is correct.

Thanks !

**Comment by AlbertSPedersen** [comment]
> Hello,
> 
> If I may bother someone to get confirmation of a testing protocol for this : would creating a session (with remember_me token) without this change, then applying the patch and keep using the session result in equivalent behavior as to what is done for AO3 deployments ? Just to make sure that what I'm testing is correct.
> 
> Thanks !

@TheCarpetMerchant testing is described [in the Jira ticket](https://otwarchive.atlassian.net/browse/AO3-6629). Existing session cookies will no longer be valid as the format is changing. The format of the "remember me" token hasn't changed, so users who selected "remember me" should automatically be issued a new session cookie without having to sign in again.

File: app/helpers/application_helper.rb [hunk 3]
```diff
@@ -191,9 +191,18 @@ def flash_div *keys
...
       if flash[key]
         if flash[key].is_a?(Array)
-          content_tag(:div, content_tag(:ul, flash[key].map { |flash_item| content_tag(:li, h(flash_item)) }.join("\n").html_safe), class: "flash #{key}")
+          content_tag(
+            :div,
+            content_tag(
+              :ul,
+              flash[key].map do |flash_item|
+                content_tag(:li, sanitize(flash_item))
+              end.join("\n").html_safe
```
Comment (weeklies) [inline app/helpers/application_helper.rb:200]: I think we can use [`safe_join`](https://apidock.com/rails/v6.1.3.1/ActionView/Helpers/OutputSafetyHelper/safe_join) to avoid html_safe.

File: app/helpers/application_helper.rb [hunk 4]
```diff
@@ -191,9 +191,18 @@ def flash_div *keys
     keys.collect { |key|
       if flash[key]
         if flash[key].is_a?(Array)
-          content_tag(:div, content_tag(:ul, flash[key].map { |flash_item| content_tag(:li, h(flash_item)) }.join("\n").html_safe), class: "flash #{key}")
+          content_tag(
+            :div,
```
Comment (weeklies) [inline app/helpers/application_helper.rb:195]: I don't think the line break before the first argument is necessary.

Comment (AlbertSPedersen) [reply app/helpers/application_helper.rb:195]: Hound was complaining about having a multi-line function call where all arguments didn't have the same indentation.

Comment (weeklies) [reply app/helpers/application_helper.rb:195]: How aboutL
```
content_tag(:div, 
            content_tag(:ul,
                        flash[key].map do |flash_item|
                          content_tag(:li, sanitize(flash_item))
                        end.join("\n").html_safe),
            class: "flash #{key}")
```

Comment (AlbertSPedersen) [reply app/helpers/application_helper.rb:195]: That's essentially what I did at first and Hound didn't like it. See https://github.com/otwcode/otwarchive/pull/4651/commits/62abf6506e400df051d76adb5183d416973b9b53. But I don't care too much how it's formatted as long as it's readable - I was just following Hound's recommendations.

Comment (AlbertSPedersen) [reply app/helpers/application_helper.rb:195]: > How about:
> 
> ```
> content_tag(:div, 
>             content_tag(:ul,
>                         flash[key].map do |flash_item|
>                           content_tag(:li, sanitize(flash_item))
>                         end.join("\n").html_safe),
>             class: "flash #{key}")
> ```

If we prefer this format, I can change it back?

Comment (sarken) [reply app/helpers/application_helper.rb:195]: Yeah, I think that's much easier to read!

File: app/helpers/application_helper.rb (hunk 3 above)
Comment (AlbertSPedersen) [reply app/helpers/application_helper.rb]: Done!

File: app/helpers/application_helper.rb (hunk 4 above)
Comment (AlbertSPedersen) [reply app/helpers/application_helper.rb]: Changed it back :)

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -98,6 +98,17 @@ def redirect_to(*args, **kwargs)
     end
   end
 
+  # Migrates session cookies from encrypted to signed format
+  before_action :migrate_session_cookie_20231230
```
[the remark this answers is not in the document]
Comment (brianjaustin) [reply app/controllers/application_controller.rb:102]: To quiet the hound (or reviewdog) you could add `# rubocop:disable Naming/VariableNumber` and `# rubocop:enable Naming/VariableNumber` around this

**Comment by sarken** [comment]
@zz9pzza was there something on this you wanted to take a look at? I noticed you [self-requested a review 3 months ago](https://github.com/otwcode/otwarchive/pull/4651#event-10820978414), so I don't want to merge this if you're still planning to take a look.

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -98,6 +98,17 @@ def redirect_to(*args, **kwargs)
...
+  def migrate_session_cookie_20231230
+    # If the request contains a valid encrypted session cookie, it means the
+    # session hasn't yet been migrated to a signed cookie, so we update the
+    # session with the contents of the encrypted cookie.
+    if cookies.encrypted[:_otwarchive_session].present?
```
[the remark this answers is not in the document]
Comment (zz9pzza) [reply app/controllers/application_controller.rb:107]: Hound is suggesting

 session.update(cookies.encrypted[:_otwarchive_session])  if cookies.encrypted[:_otwarchive_session].present?

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -191,9 +191,14 @@ def flash_div *keys
     keys.collect { |key|
       if flash[key]
         if flash[key].is_a?(Array)
-          content_tag(:div, content_tag(:ul, flash[key].map { |flash_item| content_tag(:li, h(flash_item)) }.join("\n").html_safe), class: "flash #{key}")
+          content_tag(:div,
+            content_tag(:ul,
+              safe_join(flash[key].map do |flash_item|
+                content_tag(:li, sanitize(flash_item))
```
Comment (zz9pzza) [inline app/helpers/application_helper.rb:197]: Is this related or a general clean up ?

File: config/initializers/session_store.rb (hunk 2 above)
Comment (zz9pzza) [inline config/initializers/session_store.rb:13]: Is there a reaon for ignoring rubocop ?

**Review by zz9pzza — APPROVED** [review]
